RFCs/FS-1152-For-loop-with-accumulation.md

```diff
@@ -0,0 +1,277 @@
+# F# RFC FS-1152 - For loop with accumulation
+
+The design suggestion ["for-with" syntactic sugar for folds](https://github.com/fsharp/fslang-suggestions/issues/1362) has not yet been marked "approved in principle".
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/1362)
+- [ ] Approved in principle
+- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/FILL-ME-IN)
+- [x] [Discussion](https://github.com/fsharp/fslang-design/discussions/808)
+
+# Summary
+
+A new syntax for `for`-loops that allow accumulation is introduced.
+```
+for <pat> = <expr> with <pat> in <expr> do
+    <expr>
+```
+The value of the loop body is used to update the accumulator which is returned in the end.
+
+# Motivation
+
+Currently, to do this, we need an additional mutable variable. For example,
+```fs
+let mutable sentence = "" // mutable accumulator
+for word in ["Hello"; " "; "World"; "!"] do
+    sentence <- sentence + word
+printfn "%s" sentence
+```
+
+However, the mutable variable leaks outside the loop body, which is undesirable.
+Moreover, the accumulator must be mutable - which is against functional immutable semantics.
```

Comment:
I think recursive functions should be mentioned here (I do see a brief mention in the suggestion itself, but not here).

```fsharp
let sentence =
    let rec build sentence = function
        | [] -> sentence
        | word :: words -> build (sentence + word) words

    build "" ["Hello"; " "; "World"; "!"]

printfn $"{sentence}"
```

Recursive functions can also give more control than this proposed new construct would, e.g.:

```fsharp
let sentence =
    let rec build (sentence : StringBuilder) (words : string list) =
        match words with
        | [] -> string sentence // Do something different in the last iteration.
        | word :: words -> build (sentence.Append word) words

    build (StringBuilder ()) ["Hello"; " "; "World"; "!"]

printfn $"{sentence}"
```